### DISCUSSION TITLE
unify stackable item logic

### DISCUSSION THREAD

**Discussion by scarf005** [cataclysmbn/Cataclysm-BN#4067] [General]
## Purpose of change

after #3489, there has been numerous issues and fixes with count-by-charges ("the stackable") items:

- #3973
- #4007
- #4008
- #4010
- #4012
- #4033
- #4061

however, they all does not solve the fundemental issue: they work on different APIs.

- singular items use [`game_object::detach`][detach]
- count-by-charge items use [`item::split`][split]

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/6bff212696c4832de7168d10e8b7f6db6de09e51/src/activity_actor.cpp#L1179-L1185

For every time an item may be stackable, this logic needs to be duplicated with
[`item::count_by_charges`][count_by_charges].

[count_by_charges]: <https://github.com/cataclysmbnteam/Cataclysm-BN/blob/6bff212696c4832de7168d10e8b7f6db6de09e51/src/itype.cpp#L94>
[detach]: <https://github.com/cataclysmbnteam/Cataclysm-BN/blob/6bff212696c4832de7168d10e8b7f6db6de09e51/src/game_object.cpp#L44>
[split]: <https://github.com/cataclysmbnteam/Cataclysm-BN/blob/6bff212696c4832de7168d10e8b7f6db6de09e51/src/item.cpp#L723>

This is problemetic in numerous ways:

- error-prone
- waste of time and effort
- makes code harder to read and maintain

## Proposed Change

1. Try to find functions that assume non-stackable items, e.g `detach`
2. Deprecate them and replace them with functions use assume stackable items, e.g `split`
3. Convert all non-stackable items into stackables of maximum stack size of 1.
4. _chaos ensues_
5. Fix remaining issues.

**Comment by MarcinSzymanek** [comment]
Is this being worked on?
